File: ios/Sources/Foundation/XTFWebSocket.m
//
//  XTFWebSocket.m
//  XTFoundation
//
//  Created by 崔明辉 on 2018/1/24.
//  Copyright © 2018年 Pony. All rights reserved.
//

#import "XTFWebSocket.h"
#import <SocketRocket/SRWebSocket.h>
#import "XTMemoryManager.h"

@interface XTFWebSocket()<SRWebSocketDelegate>

@property (nonatomic, copy) NSString *objectUUID;
@property (nonatomic, strong) SRWebSocket *socket;
@property (nonatomic, weak) JSContext *context;

@end

@implementation XTFWebSocket

+ (NSString *)create:(NSString *)URLString {
    XTFWebSocket *instance = [[XTFWebSocket alloc] init];
    instance.context = [JSContext currentContext];
    instance.socket = [[SRWebSocket alloc] initWithURL:[NSURL URLWithString:URLString]];
    XTManagedObject *managedObject = [[XTManagedObject alloc] initWithObject:instance];
    instance.objectUUID = managedObject.objectUUID;
    managedObject.objectThreadSafe = YES;
    [XTMemoryManager add:managedObject];
    return managedObject.objectUUID;
}

+ (void)xtr_sendData:(NSString *)dataRef objectRef:(NSString *)objectRef {
    XTFWebSocket *obj = [XTMemoryManager find:objectRef];
    NSData *data = [XTMemoryManager find:dataRef];
    if ([obj isKindOfClass:[XTFWebSocket class]] && [data isKindOfClass:[NSData class]]) {
        [obj.socket send:data];
    }
}

+ (void)xtr_sendString:(NSString *)string objectRef:(NSString *)objectRef {
    XTFWebSocket *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[XTFWebSocket class]]) {
        [obj.socket send:string];
    }
}

+ (void)xtr_close:(NSString *)objectRef {
    XTFWebSocket *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[XTFWebSocket class]]) {
        [obj.socket close];
    }
}

- (void)setSocket:(SRWebSocket *)socket {
    _socket = socket;
    socket.delegate = self;
    [socket performSelector:@selector(open) withObject:nil afterDelay:0.1];
}

- (JSValue *)scriptObject {
    return [self.context evaluateScript:[NSString stringWithFormat:@"objectRefs['%@']", self.objectUUID]];
}

- (void)webSocketDidOpen:(SRWebSocket *)webSocket {
    [[self scriptObject] invokeMethod:@"handleOpen" withArguments:@[]];
}

- (void)webSocket:(SRWebSocket *)webSocket didCloseWithCode:(NSInteger)code reason:(NSString *)reason wasClean:(BOOL)wasClean {
    [[self scriptObject] invokeMethod:@"handleClose" withArguments:@[@(code), reason ?: @""]];
}

- (void)webSocket:(SRWebSocket *)webSocket didFailWithError:(NSError *)error {
    [[self scriptObject] invokeMethod:@"handleFail" withArguments:@[error.localizedDescription ?: @""]];
}

- (void)webSocket:(SRWebSocket *)webSocket didReceiveMessage:(id)message {
    if ([message isKindOfClass:[NSData class]]) {
        XTManagedObject *managedObject = [[XTManagedObject alloc] initWithObject:message];
        managedObject.objectThreadSafe = YES;
        [XTMemoryManager add:managedObject];
        [[self scriptObject] invokeMethod:@"handleDataMessage" withArguments:@[managedObject.objectUUID]];
    }
    else if ([message isKindOfClass:[NSString class]]) {
        [[self scriptObject] invokeMethod:@"handleStringMessage" withArguments:@[message]];
    }
}

@end

File: android/foundation/src/main/java/com/opensource/xt/foundation/XTFWebSocket.kt
package com.opensource.xt.foundation

import android.os.Handler
import com.eclipsesource.v8.V8Object
import com.opensource.xt.core.*
import org.java_websocket.client.WebSocketClient
import org.java_websocket.drafts.Draft
import org.java_websocket.drafts.Draft_6455
import org.java_websocket.handshake.ServerHandshake
import java.lang.Exception
import java.net.URI
import java.nio.ByteBuffer

/**
 * Created by cuiminghui on 2018/1/30.
 */
class XTFWebSocket(val context: XTContext, URLString: String): WebSocketClient(URI.create(URLString), Draft_6455()), XTComponentInstance {

    override var objectUUID: String? = null

    init {
        Handler().postDelayed({
            connect()
        }, 100)
    }

    fun scriptObject(): V8Object? {
        return context.evaluateScript("objectRefs['$objectUUID']") as? V8Object
    }

    override fun onOpen(handshakedata: ServerHandshake?) {
        context.sharedHandler.post {
            scriptObject()?.let {
                XTContext.invokeMethod(it, "handleOpen")
                XTContext.release(it)
            }
        }
    }

    override fun onClose(code: Int, reason: String?, remote: Boolean) {
        context.sharedHandler.post {
            scriptObject()?.let {
                XTContext.invokeMethod(it, "handleClose", listOf(code, reason ?: ""))
                XTContext.release(it)
            }
        }
    }

    override fun onMessage(message: String?) {
        context.sharedHandler.post {
            scriptObject()?.let {
                XTContext.invokeMethod(it, "handleStringMessage", listOf(message ?: ""))
                XTContext.release(it)
            }
        }
    }

    override fun onMessage(bytes: ByteBuffer?) {
        val bytes = bytes ?: return
        val byteArray = ByteArray(bytes.remaining())
        bytes.get(byteArray)
        context.sharedHandler.post {
            scriptObject()?.let {
                val data = XTFData(byteArray)
                val managedObject = XTManagedObject(data)
                data.objectUUID = managedObject.objectUUID
                XTMemoryManager.add(managedObject)
                XTContext.invokeMethod(it, "handleStringMessage", listOf(managedObject.objectUUID))
                XTContext.release(it)
            }
        }
    }

    override fun onError(ex: Exception?) {
        context.sharedHandler.post {
            scriptObject()?.let {
                XTContext.invokeMethod(it, "handleFail", listOf(ex?.message ?: ""))
                XTContext.release(it)
            }
        }
    }


    class JSExports(val context: XTContext): XTComponentExport() {

        override val name: String = "_XTFWebSocket"

        override fun exports(): V8Object {
            val exports = V8Object(context.runtime)
            exports.registerJavaMethod(this, "create", "create", arrayOf(String::class.java))
            exports.registerJavaMethod(this, "xtr_sendData", "xtr_sendData", arrayOf(String::class.java, String::class.java))
            exports.registerJavaMethod(this, "xtr_sendString", "xtr_sendString", arrayOf(String::class.java, String::class.java))
            exports.registerJavaMethod(this, "xtr_close", "xtr_close", arrayOf(String::class.java))
            return exports
        }

        fun create(URLString: String): String {
            val instance = XTFWebSocket(context, URLString)
            val managedObject = XTManagedObject(instance)
            instance.objectUUID = managedObject.objectUUID
            XTMemoryManager.add(managedObject)
            return managedObject.objectUUID
        }

        fun xtr_sendData(dataRef: String, objectRef: String) {
            val data = XTMemoryManager.find(dataRef) as? XTFData ?: return
            val socket = XTMemoryManager.find(objectRef) as? XTFWebSocket ?: return
            socket.send(data.bytes)
        }

        fun xtr_sendString(stringValue: String, objectRef: String) {
            val socket = XTMemoryManager.find(objectRef) as? XTFWebSocket ?: return
            socket.send(stringValue)
        }

        fun xtr_close(objectRef: String) {
            val socket = XTMemoryManager.find(objectRef) as? XTFWebSocket ?: return
            socket.close()
        }

    }

}